system/Helpers/form_helper.php

```diff
@@ -0,0 +1,859 @@
+    /**
+     * Hidden Input Field
+     *
+     * Generates hidden fields. You can pass a simple key/value string or
+     * an associative array with multiple values.
+     *
+     * @param	mixed	$name		Field name
+     * @param	string	$value		Field value
+     * @param	bool	$recursing
+     * @return	string
+     */
+    function form_hidden($name, string $value = '', bool $recursing = false): string
+    {
+        static $form;
+        if ($recursing === false)
+        {
+            $form = "\n";
+        }
+        if (is_array($name))
+        {
+            foreach ($name as $key => $val)
+            {
+                form_hidden($key, $val, true);
+            }
+            return $form;
+        }
+        if ( ! is_array($value))
+        {
+            $form .= '<input type="hidden" name="'.$name.'" value="'.html_escape($value)."\" />\n";
+        }
+        else
+        {
+            foreach ($value as $k => $v)
+            {
+                $k = is_int($k) ? '' : $k;
+                form_hidden($name.'['.$k.']', $v, true);
+            }
+        }
+        return $form;
+    }
+}
+
+// ------------------------------------------------------------------------
+
+if ( ! function_exists('form_input'))
+{
+    /**
+     * Text Input Field
+     *
+     * @param	mixed
+     * @param	string
+     * @param	mixed
+     * @return	string
+     */
+    function form_input($data = '', string $value = '', $extra = ''): string
+    {
+        $defaults = [
+                'type' => 'text',
+                'name' => is_array($data) ? '' : $data,
+                'value' => $value
+        ];
+        return '<input '._parse_form_attributes($data, $defaults)._attributes_to_string($extra)." />\n";
+    }
+}
+
+// ------------------------------------------------------------------------
+
+if ( ! function_exists('form_password'))
+{
+    /**
+     * Password Field
+     *
+     * Identical to the input function but adds the "password" type
+     *
+     * @param	mixed
+     * @param	string
+     * @param	mixed
+     * @return	string
+     */
+    function form_password($data = '', string $value = '', $extra = ''): string
+    {
+        is_array($data) OR $data = array('name' => $data);
+        $data['type'] = 'password';
+        return form_input($data, $value, $extra);
+    }
+}
+
+// ------------------------------------------------------------------------
+
+if ( ! function_exists('form_upload'))
+{
+    /**
+     * Upload Field
+     *
+     * Identical to the input function but adds the "file" type
+     *
+     * @param	mixed
+     * @param	string
+     * @param	mixed
+     * @return	string
+     */
+    function form_upload($data = '', string $value = '', $extra = ''): string
+    {
+        $defaults = ['type' => 'file', 'name' => ''];
+        is_array($data) OR $data = ['name' => $data];
+        $data['type'] = 'file';
+        return '<input '._parse_form_attributes($data, $defaults)._attributes_to_string($extra)." />\n";
```

Comment:
Please include blank lines between logical groups within code for better scanability/readability.